### PR TITLE
Tests for fetchProfileDiffUnobfuscated and filtering out id in addOrUpdate

### PR DESCRIPTION
**Pull request by RahulGoyal-tech**
Date: 18-11-2024

Developer Name: Rahul Goyal

---

## Issue Ticket Number

#2223 
## Description

Added tests for the PR #2253 
- Added unit tests for `profileDiffs.js` model to test `fetchProfileDiffUnobfuscated`.
- Tested it to give unobfuscated phone and mail
- Tested it to give `{ profileDiffExists : false }` if invalid diff id.
- Tested it to give a valid id and `{ profileDiffExists : true }` if given valid diff id.
- Added test to ensure if id field is filtered out while updating diffs

### Documentation Updated?

- [ ] Yes
- [x] No

### Under Feature Flag

- [ ] Yes
- [x] No

### Database Changes

- [ ] Yes
- [x] No

### Breaking Changes

- [ ] Yes
- [x] No

### Development Tested?

- [x] Yes
- [ ] No

## Test Coverage
<details>
<summary>Screenshot 1</summary>

image

image

image

</details>

### CODE REVIEW

**Comment by RahulGoyal-tech** [comment]
Stacked PR Containing test for #2253

File: models/users.js [hunk 1]
```diff
@@ -42,9 +42,17 @@ const addOrUpdate = async (userData, userId = null) => {
...
       // user exists update user
       if (user.data()) {
+        // remove id field if exist in fetched data or profileDiff
+        const currentData = user.data();
+        if ("id" in currentData) {
+          delete currentData.id;
+        }
+        if ("id" in userData) {
+          delete userData.id;
+        }
```
Comment (prakashchoudhary07) [inline models/users.js:52]: Help me understand, why is this being done?

File: test/unit/models/profileDiffs.test.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
+/**
+ * This eslint rule is disabled because of https://github.com/nodesecurity/eslint-plugin-security/issues/21
+ * It gives linting errors in testing the DB data with keys from fixtures
+ */
+/* eslint-disable security/detect-object-injection */
```
Comment (prakashchoudhary07) [inline test/unit/models/profileDiffs.test.js:5]: Please solve the issues, find the reason why its giving lint error and try solving it

File: test/unit/models/profileDiffs.test.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+      const diff = await profileDiffsQuery.fetchProfileDiffUnobfuscated(profileDiffId);
+      expect(diff.profileDiffExists).to.equal(false);
+    });
+
+    it("should return profileDiffs with unobfuscated email and phone", async function () {
+      const profileDiffs = getProfileDiffs();
+      const profileDiffId = await profileDiffsQuery.add(profileDiffs[0]);
+      const diff = await profileDiffsQuery.fetchProfileDiffUnobfuscated(profileDiffId);
+      expect(diff.phone).to.match(/^(\+\d{1,3}[- ]?)?\d{7,10}$/);
+      expect(diff.email).to.match(/^([A-Z0-9+-]{1,20}(.[A-Z0-9+-]{1,20})*)@[A-Z0-9][A-Z0-9-](.[A-Z]{2,})+$/i);
```
Comment (prakashchoudhary07) [inline test/unit/models/profileDiffs.test.js:40]: So since fixtures are used here, please match the output and input. Instead of using a regex

File: models/users.js (hunk 1 above)
Comment (RahulGoyal-tech) [reply models/users.js:52]: In the current implementaton, user.data() is the present data of user being fetched and userData are the profileDiffs. In this section, I am checking if an id field in present in document data and then deleting it. As we need to filter out id field while pushing diffs.

Note: In line 59, we are using { merge : true } so we actually do not really need to fetch and push user.data() at all or work on currentData but it is there in the present code so i updated it accordingly. Discussed with lakshay to make a seperate issue for removal of same.

File: test/unit/models/profileDiffs.test.js (hunk 2 above)
Comment (RahulGoyal-tech) [reply test/unit/models/profileDiffs.test.js:5]: Ohh alright. Will resolve this.

File: test/unit/models/profileDiffs.test.js (hunk 3 above)
Comment (RahulGoyal-tech) [reply test/unit/models/profileDiffs.test.js:40]: Sure, will update.

Comment (RahulGoyal-tech) [reply test/unit/models/profileDiffs.test.js]: @prakashchoudhary07 Updated

File: test/unit/models/profileDiffs.test.js (hunk 2 above)
Comment (RahulGoyal-tech) [reply test/unit/models/profileDiffs.test.js]: @prakashchoudhary07 Resolved. No linting issue now.

**Comment by RahulGoyal-tech** [comment]
image

File: models/users.js [hunk 4]
```diff
@@ -42,9 +42,17 @@ const addOrUpdate = async (userData, userId = null) => {
       const isNewUser = !user.data();
       // user exists update user
       if (user.data()) {
+        // remove id field if exist in fetched data or profileDiff
+        const currentData = user.data();
```
Comment (prakashchoudhary07) [inline models/users.js:46]: Does this has id in it?

Comment (prakashchoudhary07) [reply models/users.js:46]: Will this ever get id? according to our model

Comment (RahulGoyal-tech) [reply models/users.js:46]: This PR is only for tests. Will be updating this after merging that and then only tests would be present and then would be merging.

Comment (prakashchoudhary07) [reply models/users.js:46]: Then do a rebase of it?

File: test/unit/models/profileDiffs.test.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+      expect(diff.profileDiffExists).to.equal(false);
+    });
+
+    it("should return profileDiffs with unobfuscated email and phone", async function () {
+      const profileDiffs = getProfileDiffs();
+      const profileDiffId = await profileDiffsQuery.add(profileDiffs[0]);
+      const diff = await profileDiffsQuery.fetchProfileDiffUnobfuscated(profileDiffId);
+      expect(diff.phone).to.equal(profileDiffs[0].phone);
+      expect(diff.email).to.equal(profileDiffs[0].email);
+    });
```
Comment (prakashchoudhary07) [inline test/unit/models/profileDiffs.test.js:35]: Did we handle all the edge cases? or do we need more tests?

File: models/users.js (hunk 4 above)
Comment (RahulGoyal-tech) [reply models/users.js:46]: Yes

File: test/unit/models/profileDiffs.test.js (hunk 5 above)
Comment (RahulGoyal-tech) [reply test/unit/models/profileDiffs.test.js:35]: I thought about cases and tried perplexity as well. Can't think of more. Can you please suggest some if required.

File: test/unit/models/profileDiffs.test.js [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+  describe("fetchProfileDiffUnobfuscated", function () {
+    it("should successfully return profileDiffs", async function () {
+      const profileDiffs = getProfileDiffs();
+      const profileDiffId = await profileDiffsQuery.add(profileDiffs[1]);
+      const diff = await profileDiffsQuery.fetchProfileDiffUnobfuscated(profileDiffId);
+      expect(diff).haveOwnProperty("id");
+      expect(diff.profileDiffExists).to.equal(true);
+    });
+
```
Comment (Achintya-Chatterjee) [inline test/unit/models/profileDiffs.test.js:20]: we should test the behavior when `profileDiffId` is `null`, `undefined`, or an empty string.

File: test/unit/models/profileDiffs.test.js [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    it("should not return profileDiffs for invalid id", async function () {
```
Comment (Achintya-Chatterjee) [inline test/unit/models/profileDiffs.test.js:21]: we should write test where the data returned from the database has unexpected types or missing fields.

File: test/unit/models/profileDiffs.test.js [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+      const profileDiffs = getProfileDiffs();
```
Comment (Achintya-Chatterjee) [inline test/unit/models/profileDiffs.test.js:22]: we should include email and phone fields with special characters, unusual formatting, or encoding issues.

File: test/unit/models/profileDiffs.test.js [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+      await profileDiffsQuery.add(profileDiffs[0]);
+      const profileDiffId = "invalid-id";
+      const diff = await profileDiffsQuery.fetchProfileDiffUnobfuscated(profileDiffId);
```
Comment (Achintya-Chatterjee) [inline test/unit/models/profileDiffs.test.js:25]: we should test with unusually large or deeply nested profileDiff objects to ensure performance and stability.

File: test/unit/models/users.test.js [hunk 10]
```diff
@@ -115,6 +115,16 @@ describe("users", function () {
       const { user } = await users.fetchUser({ githubUsername });
       expect(user).to.haveOwnProperty("github_created_at");
     });
+
+    it("it should filter out the id field while updating profileDiff", async function () {
+      const userData = userDataArray[0];
+      const { userId } = await users.addOrUpdate(userData);
+      const profileDiff = { id: "random-id", diff: "random-diff" };
+      await users.addOrUpdate(profileDiff, userId);
```
Comment (Achintya-Chatterjee) [inline test/unit/models/users.test.js:123]: we should verify how overlapping keys between user.data() and profileDiff are resolved.

File: test/unit/models/users.test.js [hunk 11]
```diff
@@ -115,6 +115,16 @@ describe("users", function () {
       const { user } = await users.fetchUser({ githubUsername });
       expect(user).to.haveOwnProperty("github_created_at");
     });
+
+    it("it should filter out the id field while updating profileDiff", async function () {
+      const userData = userDataArray[0];
+      const { userId } = await users.addOrUpdate(userData);
+      const profileDiff = { id: "random-id", diff: "random-diff" };
```
Comment (Achintya-Chatterjee) [inline test/unit/models/users.test.js:122]: we should test with partial data where only a subset of fields in the user profile is updated.

File: test/unit/models/users.test.js [hunk 12]
```diff
@@ -115,6 +115,16 @@ describe("users", function () {
       const { user } = await users.fetchUser({ githubUsername });
       expect(user).to.haveOwnProperty("github_created_at");
     });
+
+    it("it should filter out the id field while updating profileDiff", async function () {
```
Comment (Achintya-Chatterjee) [inline test/unit/models/users.test.js:119]: we should test with malformed user data (e.g., missing `githubUsername` `or non-JSON serializable fields).

Comment (Achintya-Chatterjee) [inline test/unit/models/users.test.js:119]: we should write test cases where the `devFeatureFlag` query parameter is missing or incorrectly set.

File: test/unit/models/users.test.js [hunk 13]
```diff
@@ -115,6 +115,16 @@ describe("users", function () {
...
+      const userData = userDataArray[0];
+      const { userId } = await users.addOrUpdate(userData);
```
Comment (Achintya-Chatterjee) [inline test/unit/models/users.test.js:121]: we should write test cases with an invalid or missing message field in the request body.

File: test/unit/models/users.test.js [hunk 14]
```diff
@@ -115,6 +115,16 @@ describe("users", function () {
...
+      const profileDiff = { id: "random-id", diff: "random-diff" };
+      await users.addOrUpdate(profileDiff, userId);
+      const data = (await userModel.doc(userId).get()).data();
+      expect(data).to.haveOwnProperty("diff"); // Ensures profile diffs are pushed
```
Comment (Achintya-Chatterjee) [inline test/unit/models/users.test.js:125]: Do you think comment is necessary here?

File: test/unit/models/users.test.js [hunk 15]
```diff
@@ -115,6 +115,16 @@ describe("users", function () {
...
+      expect(data.id).not.equal("random-id"); // Ensures id is not added / updated by profile Diffs
```
Comment (Achintya-Chatterjee) [inline test/unit/models/users.test.js:126]: Do you think comment is necessary here?

File: test/unit/models/users.test.js (hunk 12 above)
Comment (Achintya-Chatterjee) [inline test/unit/models/users.test.js:119]: we should test scenarios where no matching `profileDiffId` is found in the database.

**Review by Achintya-Chatterjee — CHANGES_REQUESTED** [review]

File: test/unit/models/users.test.js (hunk 10 above)
Comment (RahulGoyal-tech) [reply test/unit/models/users.test.js:123]: I think this had to be resolved because we are removing PR #2282. We had a discussion around this.

File: test/unit/models/users.test.js (hunk 11 above)
Comment (RahulGoyal-tech) [reply test/unit/models/users.test.js:122]: Doing this in Line 119 - Line 127 where only 2 fields are updated using profileDiff

**Review by Achintya-Chatterjee — APPROVED** [review]

**Review by MehulKChaudhari — APPROVED** [review]

**Comment by vikasosmium** [comment]
this should be in one pr #2253

**Comment by RahulGoyal-tech** [comment]
> this should be in one pr #2253

Had a conversation in beginning with achintya. Planned at that time to make this stacked.

**Comment by vikasosmium** [comment]
> > this should be in one pr #2253
> 
> Had a conversation in beginning with achintya. Planned at that time to make this stacked.

I don't think it needed stacked pr here.

**Comment by RahulGoyal-tech** [comment]
> > > this should be in one pr #2253
> > 
> > 
> > Had a conversation in beginning with achintya. Planned at that time to make this stacked.
> 
> I don't think it needed stacked pr here.

Yeah it could be encapsulated in one pr itself as well. Please review and approve this. Will try to keep single pr for stacked ones from next time.
